source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -4351,6 +4351,148 @@ export class ColumnSet extends CardElementContainer {
     }
 }
 
+export class Source {
+    mimeType: string;
+    url: string;
+
+    parse(json: any): Source {
+        this.mimeType = json["mimeType"];
+        this.url = json["url"];
+        return this;
+    }
+
+    toJSON() {
+        return { mimeType: this.mimeType, url: this.url };
+    }
+}
+
+export class Media extends CardElement {
+    poster: string;
+    altText: string;
+    sources: Array<Source> = [];
+    private _mediaType: string;
+    private _mediaElement: HTMLMediaElement;
+    static readonly acceptedMediaTypes = ["audio", "video"];
+
+    private processSources(): Array<Source> {
+        let typeFilter = (t): boolean => {
+            if (Media.acceptedMediaTypes.indexOf(t)) {
+                this._mediaType = t;
+                typeFilter = (tt) => tt === this._mediaType;
+                return true;
+            }
+            return false;
+        }
+        return this.sources.filter((s): boolean => {
+            if (s.mimeType) {
+                let mimeComponents = s.mimeType.split('/');
+                if (mimeComponents.length == 2) {
+                    return typeFilter(mimeComponents[0]);
+                }
+            }
+            return false;
+        });
+    }
+
+    private makeButton(label: string): HTMLAnchorElement {
+        let button = document.createElement("a");
+        button.className = "ac-media-button";
+        button.text = label;
+        return button;
+    }
+
+    protected internalRender(): HTMLElement {
+        var element = document.createElement("div");
+
+        element.className = this.hostConfig.makeCssClassName("ac-media");
+
+        let goodSources = this.processSources();
+
+        if (goodSources.length) {
+
+            this._mediaElement = <HTMLMediaElement>document.createElement(this._mediaType);
+            this._mediaElement.style.width = "100%";
+
+            goodSources.forEach(s => {
+                let src: HTMLSourceElement = document.createElement("source");
+                src.src = s.url;
+                src.type = s.mimeType;
+                this._mediaElement.appendChild(src);
+            });
+            element.appendChild(this._mediaElement);
+
+            var play_button = this.makeButton("▶️");
+            play_button.onclick = e => this.play();
+            element.appendChild(play_button);
+
+            var pause_button = this.makeButton("⏸️");
+            pause_button.onclick = e => this.pause();
+            element.appendChild(pause_button);
+
+            var stop_button = this.makeButton("⏹️");
+            stop_button.onclick = e => this.stop();
+            element.appendChild(stop_button);
+        }
+
+        return element;
+    }
+
+    play() {
+        this._mediaElement.play();
```

Comment:
We probably want to hide the play button and show the pause button instead

Reply:
This is the kind of feedback I was after at this time, thanks for it!